File: CustomView/app/src/main/java/com/ohshell/customview/MainActivity.java
package com.ohshell.customview;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        findViewById(R.id.btn_number_input).setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        int id = v.getId();
        if (id == R.id.btn_number_input) {
            Intent intent = new Intent(this, NumberInputActivity.class);
            startActivity(intent);
        }
    }
}

File: testprovider/app/src/main/java/com/owulia/testprovider/MainActivity.kt
package com.owulia.testprovider

import android.content.ContentValues
import android.database.ContentObserver
import android.net.Uri
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.Handler
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val uri = Uri.parse("content://com.owulia.mvvmdemo/user")
        contentResolver.registerContentObserver(uri, true, object : ContentObserver(Handler()){
            override fun onChange(selfChange: Boolean) {
                super.onChange(selfChange)
                println("数据发生了变化")
            }
        })


        /**
         * 获取数据
         * */
        vButton.setOnClickListener {
            val cursor = contentResolver.query(uri, null, null, null, null, null)
            val columnNames = cursor?.columnNames
            while (cursor?.moveToNext() == true) {
                columnNames?.forEach {
                    val value = cursor.getString(cursor.getColumnIndex(it))
                    println("name => $it value => $value")
                }
            }
            cursor?.close()
        }

        /**
         * 插入数据
         * */
        vAddButton.setOnClickListener {
            val values = ContentValues().apply {
                put("userName", "陈1")
                put("sex", "nan")
                put("age", 1)
                put("password", "123456")
            }
            contentResolver.insert(uri, values)
        }
    }
}
